### PR TITLE
[distributed] Add inductor-ready SNN distributed training stack

### PR DESCRIPTION
**Pull request by fangwei123456**
## Summary
- add an experimental DTensor/FSDP2/TP/PP distributed training stack for activation-based SNNs
- add standard neuron multi-step `backend="inductor"` support and integrate it with distributed benchmarks/examples
- extend distributed benchmarks, memopt integration, strategy recommendation, and tutorials with measured results

## Validation
- local pytest for distributed, benchmark, compile-compat, spikformer, and memopt coverage
- server-side CUDA smoke and benchmark validation for `dp`, `tp`, `fsdp2`, `fsdp2_tp`, and `pp`

## Summary by CodeRabbit

* **New Features**
  * Experimental distributed SNN training toolchain and CLI, distributed benchmarking with automated JSONL comparisons, a new Spikformer model family, and additional FlexSN benchmarking utilities.

* **Documentation**
  * New English/Chinese distributed-training tutorials and API docs; tutorials added to site navigation.

* **Performance / Tools**
  * Broader inductor backend support, FlexSN final-state and segmented execution, and memopt plan caching for faster repeated optimization.

* **Tests**
  * Large expansion of tests covering distributed, inductor, FlexSN, Spikformer, memopt, and benchmarking.

* **Bug Fixes**
  * Robust spike decompression shape handling and attention head-dimension validation.

### CODE REVIEW

**Comment by fangwei123456** [comment]
@coderabbitai review

@gemini-code-assist review

@codex review

**Comment by fangwei123456** [comment]
This PR supersedes #666 and will be the single active review thread for the latest distributed/inductor work.

**Comment by fangwei123456** [comment]
Pushed follow-up fixes for the latest review findings.

@coderabbitai review

@gemini-code-assist review

@codex review

**Comment by fangwei123456** [comment]
Pushed follow-up fixes for the latest review findings in `d494e1b6`.

This round fixes:
- pipeline schedule validation for `gpipe`/`1f1b` with `pp_virtual_stages > 1`
- pipeline module construction when the first logical stage is empty (`transpose_input` / `patch_embed` now attach to the first non-empty stage)
- batch-major classification outputs in both the distributed benchmark and distributed training script
- stale memopt plan cache replay fallback
- matching regression tests for the cases above

Validated locally with:
- `pytest test/activation_based/test_benchmark_snn_distributed.py test/activation_based/test_distributed_dtensor.py test/activation_based/test_checkpointing.py -q`
- `py_compile` on the touched distributed/memopt scripts

@coderabbitai review

@gemini-code-assist review

@codex review

**Comment by fangwei123456** [comment]
Pushed another follow-up in `4eb4e8ca`.

This round fixes:
- non-pipeline benchmark now synchronizes across ranks and reports max elapsed / max peak memory
- pipeline training only moves images to the first stage and labels to the last stage
- distributed teardown no longer adds an unconditional barrier before `destroy_process_group()`
- memopt plan-cache keys now include the profiling device
- added ZeRO-path and multi-step output regression coverage
- updated the FlexSN benchmark to fail fast when `PYTORCH_JIT != 0`

Validated locally with:
- `pytest test/activation_based/test_benchmark_snn_distributed.py test/activation_based/test_distributed_dtensor.py test/activation_based/test_checkpointing.py -q`
- `py_compile` on the touched benchmark/train/memopt files

@coderabbitai review

@gemini-code-assist review

@codex review

**Comment by fangwei123456** [comment]
Pushed `d817516c` to `codex/dtensor-snn-distributed`.

This round fixes the remaining actionable review items I found:
- reset stage-local SNN state after each pipeline benchmark step
- synchronize CUDA before timing `benchmark_pipeline()` elapsed wall time
- strengthen `reduce_classification_output` tests to assert values, not just shapes
- restore the global memopt plan cache after cache-isolation tests
- clean up the raw regex / unused local test nits

Validation rerun:
- `pytest test/activation_based/test_benchmark_snn_distributed.py test/activation_based/test_distributed_dtensor.py test/activation_based/test_checkpointing.py -q`
- `60 passed, 2 skipped`
- `py_compile` on touched files passed

@coderabbitai review

@gemini-code-assist review

@codex review

**Comment by fangwei123456** [comment]
Pushed `1449b10a` to `codex/dtensor-snn-distributed`.

This round tightens the remaining benchmark/model issues from review:
- include more benchmark-shaping knobs in the JSONL comparison key and persisted records
- only allocate PP inputs/labels on the stages that actually consume them
- strengthen the empty-first-stage pipeline tests with real baseline-vs-pipeline forward assertions
- switch `activation_based.model.__init__` to explicit Spikformer re-exports with `__all__`
- add public Spikformer factory return annotations and exercise the package re-export in tests
- restore per-iteration state reset in `benchmark/flexsn/flex_sn_inductor.py` and standardize the config-error exit code

Validation rerun:
- `pytest test/activation_based/test_benchmark_snn_distributed.py test/activation_based/test_distributed_dtensor.py test/activation_based/test_checkpointing.py test/activation_based/test_spikformer_model.py -q`
- `62 passed, 2 skipped`
- `py_compile` on touched files passed

@coderabbitai review

@gemini-code-assist review

@codex review

**Comment by fangwei123456** [comment]
Pushed `25aa30e2` to `codex/dtensor-snn-distributed`.

This round addresses the remaining real pipeline/benchmark issues I saw after the last review sweep:
- keep PP ranks on the same sample order by giving `train_distributed.build_data()` a shared sampler path for pipeline mode
- stream JSONL comparison reads instead of loading the entire benchmark history into memory
- validate benchmark CLI positive-only knobs up front (`batch-size`, `T`, `steps`, `image-size`, `num-classes`, `pp-microbatches`, `pp-virtual-stages`)
- sort the new Spikformer `__all__` exports
- bind loop-scoped benchmark lambdas safely in `benchmark/flexsn/flex_sn_inductor.py`
- align the PYTORCH_JIT hard-error text with the script’s actual failure behavior and header usage

Validation rerun:
- `pytest test/activation_based/test_benchmark_snn_distributed.py test/activation_based/test_distributed_dtensor.py test/activation_based/test_checkpointing.py test/activation_based/test_spikformer_model.py -q`
- `64 passed, 2 skipped`
- `py_compile` on touched files passed

@coderabbitai review

@gemini-code-assist review

@codex review

**Comment by fangwei123456** [comment]
Pushed `95498136` (`Fix remaining distributed review blockers`).

Addressed the remaining actionable review items:
- reset pipeline stage state after each `optimizer.step()` in `train_one_epoch_pipeline()`
- reject 1D `--mesh-shape` in `fsdp2_tp` benchmark mode
- remove benchmark cleanup barriers before `destroy_process_group()`
- synchronize CUDA before the PP timing barrier
- strengthen the shared pipeline sampler test to assert `num_replicas == 1` and `rank == 0`

Validation:
- `pytest test/activation_based/test_distributed_dtensor.py test/activation_based/test_benchmark_snn_distributed.py -q`
- `49 passed, 2 skipped`
- `py_compile` passed for the touched benchmark/train/test files

@coderabbitai review

@gemini-code-assist review

@codex review
